File: javascript/Day05-Exer01.js
// Permutations: Hoán vị

function allPermutation(nums){
    //write code here
    let result = [];
    let temp = [];
    let visited = [];
    if (nums.length == 0) return result;
    if (nums.length == 1) return [nums];
    for(let i = 0; i < nums.length; i++){
        visited[i] = false;
    }
    backTracking(result, temp, nums, visited);
    return result;
};

function backTracking(result, temp, nums, visited){
    if(temp.length == nums.length){
        result.push([...temp]);
        return;
    }
    for(let i = 0; i < nums.length; i++){
        if(visited[i] == false){
            visited[i] = true;
            temp.push(nums[i]);
            backTracking(result, temp, nums, visited);
            temp.pop();
            visited[i] = false;
        }
    }
};

// Solution 2: Using swap and recursion
// Time complexity: O(n!)
function allPermutation(nums){
    if(nums.length ==0) return [[]];
    const permutation = [];
    function helper(nums,i){
        if(i=== nums.length-1){
            permutation.push(nums.slice());
            return;
        }
        for(let j=i;j<nums.length;j++){
            //swap i,j
            [nums[i],nums[j]] = [nums[j],nums[i]];
            //recursive
            helper(nums,i+1);
            //swap i,j 
            [nums[i],nums[j]] = [nums[j],nums[i]];
        }
    }
    helper(nums,0);
    return permutation;
};

File: python/Day05-Exer01.py
# Permutations: Hoán vị

def all_permutations(nums):
    #Write code here
    result = []
    if len(nums) == 0:
        return result
    if len(nums) == 1:
        return [nums]
    for i in range(len(nums)):
        m = nums[i]
        remain = nums[:i] + nums[i+1:]
        for p in all_permutations(remain):
            result.append([m] + p)
    return result

# Solution
def all_permutations(nums):
    permutations = []
    if len(nums)==0: return [[]]
    def helper(nums,i):
        if i == len(nums) -1:
            permutations.append(nums.copy())
            return
        for j in range(i,len(nums)):
            nums[i],nums[j] = nums[j],nums[i]
            helper(nums,i+1)
            nums[i],nums[j] = nums[j],nums[i]
    helper(nums,0)
    return permutations


array = [1,2,3]
print(all_permutations(array))


